### PR TITLE
Introduce the `NoCalcLength`

### PR DESCRIPTION
**Pull request by ghost**
I began this for making the `CalcLengthOrPercentage` represent `LengthOrPercentage` (instead of the enum we already have), but only later did I realize that it will make `LengthOrPercentageOrFoo` types fatty (which is the problem we're trying to avoid - #15061) and so, I dropped that attempt. Along the way, I introduced an internal type for `Length`, for representing all its non-calc variants (which are `Copy`). We could still have this type for the `LengthOrPercentageOrFoo` types which don't really need  `Length` since they already have their own variants for calc.

r? @Manishearth @emilio @SimonSapin or anyone interested

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because it's a refactor

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15115)

### CODE REVIEW

File: components/style/properties/longhand/box.mako.rs [hunk 1]
```diff
@@ -1344,7 +1344,7 @@ ${helpers.single_keyword("animation-fill-mode",
                             TranslateKind::TranslateY,
                             specified::LengthOrPercentage::zero(),
                             ty,
-                            specified::Length::Absolute(Au(0))));
+                            specified::Length::default()));
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: What about `zero` instead of `default`?

**Comment by Wafflespeanut** [comment]
Hm, looks like I forgot the geckolib stuff.

File: components/style/properties/longhand/box.mako.rs (hunk 1 above)
Comment (Wafflespeanut) [inline components/style/properties/longhand/box.mako.rs]: Agreed. That's what we have in `LoP`

**Comment by emilio** [comment]
Beware for the unit tests too

**Comment by Manishearth** [comment]
Can you make all of the `LengthOrFoo` types use LengthInner as well? They already have separate calc variants

**Comment by Wafflespeanut** [comment]
Yes, that's the plan, but I thought it could be a bunch of easy issues?

**Comment by SimonSapin** [comment]
Small nitpick: could you rename the `Length::Inner` variant and `LengthInternal` to something more descriptive? Maybe `NonCalc`/`NonCalcLength`.

**Comment by Manishearth** [comment]
> Yes, that's the plan, but I thought it could be a bunch of easy issues?

I'm suspecting we'll have some tricky issues with getting the tests to work again even though it's a refactor. idk.

**Comment by Wafflespeanut** [comment]
@SimonSapin @Manishearth Okay, will do.

**Comment by Wafflespeanut** [comment]
This is ready for review.

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -455,18 +454,18 @@ impl LayoutElementHelpers for LayoutJS<Element> {
...
 
-        let font_size = if let Some(this) = self.downcast::<HTMLFontElement>() {
-            this.get_size()
-        } else {
-            None
-        };
+        let font_size = self.downcast::<HTMLFontElement>().and_then(|this| {
+            this.get_size().map(|len| match len {
+                specified::Length::NoCalc(l) => LengthOrPercentage::Length(l),
+                specified::Length::Calc(l, _) => LengthOrPercentage::Calc(l),
```
Comment (emilio) [inline components/script/dom/element.rs]: Probably worth a method in `Length`, `Length::into_length_or_percentage`?

File: components/style/values/computed/length.rs [hunk 3]
```diff
@@ -17,6 +17,45 @@ pub use super::image::{EndingShape as GradientShape, Gradient, GradientKind, Ima
...
+    }
+
+    #[inline]
+    fn from_computed_value(computed: &Au) -> Self {
+        specified::NoCalcLength::Absolute(*computed)
+    }
+}
+
+impl ToComputedValue for specified::Length {
+    type ComputedValue = Au;
```
Comment (emilio) [inline components/style/values/computed/length.rs]: The computed value of a length can also be a `Calc`, as far as I know, right?

File: components/style/values/specified/basic_shape.rs [hunk 4]
```diff
@@ -327,7 +327,7 @@ fn serialize_basicshape_position<W>(position: &Position, dest: &mut W)
         // 0 length should be replaced with 0%
         fn replace_with_percent(input: LengthOrPercentage) -> LengthOrPercentage {
             match input {
-                LengthOrPercentage::Length(Length::Absolute(au)) if au.0 == 0 => {
+                LengthOrPercentage::Length(ref l) if l == &NoCalcLength::zero() => {
```
Comment (emilio) [inline components/style/values/specified/basic_shape.rs]: probably worth adding a method to `NoCalcLength` (`NoCalcLength::is_zero`?)

File: components/style/values/specified/length.rs [hunk 5]
```diff
@@ -205,12 +213,10 @@ impl CharacterWidth {
     }
 }
 
-/// A length.
-///
-/// https://drafts.csswg.org/css-values/#lengths
-#[derive(Clone, PartialEq, Debug)]
+/// A length that cannot parse `calc` expressions
```
Comment (emilio) [inline components/style/values/specified/length.rs]: A `<length>`, without taking `calc` into account (and with spec links to both sections) would be the ideal IMO.

File: components/style/values/specified/length.rs [hunk 6]
```diff
@@ -229,7 +235,107 @@ pub enum Length {
...
+    }
+
+    /// Get an absolute length from a px value.
+    #[inline]
+    pub fn from_px(px_value: CSSFloat) -> NoCalcLength {
+        NoCalcLength::Absolute(Au((px_value * AU_PER_PX) as i32))
+    }
+}
 
+/// A length
```
Comment (emilio) [inline components/style/values/specified/length.rs]: A `<length>`, extended by `calc()` (again, with links). Extended may not be the best wording, so feel free to tweak as appropriate.

**Review by emilio — COMMENTED** [review]
This looks way cleaner IMO! Thanks for taking the time to do this :)

I have mostly minor suggestions.

File: components/style/values/computed/length.rs (hunk 3 above)
Comment (emilio) [inline components/style/values/computed/length.rs]: This is pre-existing, so probably that's ok.

Comment (Wafflespeanut) [inline components/style/values/computed/length.rs]: I'll take a look at this later :)

**Comment by Wafflespeanut** [comment]
Addressed all the comments. r? @emilio

**Comment by emilio** [comment]
Tidy is failing:

```
./tests/unit/style/viewport.rs:16: use statement is not in alphabetical order
	expected: style::values::specified::NoCalcLength::{self, ViewportPercentage}
	found: style::values::specified::LengthOrPercentageOrAuto::{self, Auto}
```

Also, please no "Fix unit tests" commit, please do squash.

Thanks for fixing this @Wafflespeanut! This looks good to me with the above fixed :)

**Comment by Wafflespeanut** [comment]
Fixed tidy failure and squashed the test. Thanks :)

@bors-servo r=emilio

**Comment by Wafflespeanut** [comment]
Meh, warning...

@bors-servo r=emilio

**Pull request merged**